### PR TITLE
Make goto-definition go to a signature declaration if possible

### LINKED ISSUE #9269: Goto definition for `super()` constructor call should goto the constructor, not the class
Given:

    class A {
        constructor() {}
    }
    class B extends A {
        constructor() {
            super();
        }
    }

I expect goto-definition for `super()` to go to `constructor() {}` and not to `class A`.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #9269 and #10074

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2788,18 +2788,37 @@ namespace ts {
...
+    function tryGetCalledDeclaration(typeChecker: TypeChecker, node: Node): SignatureDeclaration | undefined {
+        const callOrNewExpression = getCallOrNewExpressionTargetingNode(node);
+        if (callOrNewExpression) {
+            const signature = typeChecker.getResolvedSignature(callOrNewExpression);
+            return signature.declaration;
         }
-        return node && node.parent && node.parent.kind === SyntaxKind.NewExpression && (<CallExpression>node.parent).expression === node;
+    }
+
+    function getCallOrNewExpressionWorker(node: Node, kind: SyntaxKind): Node | undefined {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: Change the type of `kind` to `SyntaxKind.CallExpression | SyntaxKind.NewExpression`

File: src/services/services.ts [hunk 2]
```diff
@@ -2788,18 +2788,37 @@ namespace ts {
...
+        return !!getCallOrNewExpressionWorker(node, SyntaxKind.CallExpression);
     }
 
     function isNewExpressionTarget(node: Node): boolean {
-        if (isRightSideOfPropertyAccess(node)) {
-            node = node.parent;
+        return !!getCallOrNewExpressionWorker(node, SyntaxKind.NewExpression);
+    }
+
+    function getCallOrNewExpressionTargetingNode(node: Node): CallExpression | NewExpression | undefined {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: `getCallOrNewTargetAncestor`

File: src/services/services.ts [hunk 3]
```diff
@@ -2788,18 +2788,37 @@ namespace ts {
...
+
     function isCallExpressionTarget(node: Node): boolean {
-        if (isRightSideOfPropertyAccess(node)) {
-            node = node.parent;
-        }
-        return node && node.parent && node.parent.kind === SyntaxKind.CallExpression && (<CallExpression>node.parent).expression === node;
+        return !!getCallOrNewExpressionWorker(node, SyntaxKind.CallExpression);
     }
 
     function isNewExpressionTarget(node: Node): boolean {
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: The semantics of these functions are different now. They no longer only check if their operand is the target of the call. Instead, they climb up from any given member. Is this behavior intended?

If you don't need to change their behavior, I'd rather you just made `getCallOrNewExpressionWorker` take a single node. Then make `getCallOrNewTargetAncestor` pass it the result of calling `climbPastPropertyAccess`.

That way, you also save two climbs (one for `CallExpression` and one for `NewExpression`)

**Comment by DanielRosenwasser** [comment]
For the record, calls and `new`s are not the only things that undergo overload resolution. Tagged templates and decorators do as well.

If it's not too hard, it'd be ideal to keep the behavior consistent between them.

**Comment by DanielRosenwasser** [comment]
Can you also add a test for going to definition on a call where overload resolution failed?

```ts
declare function doStuff(n: number): number;
declare function doStuff(s: string): string;

/**/doStuff(true);
```

Ideally, the language service should still jump to one of the overloads.

File: src/services/services.ts (hunk 1 above)
Comment (andy-ms) [inline src/services/services.ts]: We'd need to do a new before using enums-as-unions.

File: src/services/services.ts [hunk 4]
```diff
@@ -2788,20 +2788,36 @@ namespace ts {
...
-        }
+        node = climbPastPropertyAccess(node);
         return node && node.parent && node.parent.kind === SyntaxKind.NewExpression && (<CallExpression>node.parent).expression === node;
     }
 
+    /** Returns a CallLikeExpression where `node` is the target being invoked. */
+    function getAncestorCallLikeExpression(node: Node): CallLikeExpression | undefined {
+        const target = climbPastManyPropertyAccesses(node);
+        const callLike = target.parent;
+        return isCallLikeExpression(callLike) && getInvokedExpression(callLike) === target && callLike;
```
Comment (zhengbli) [inline src/services/services.ts]: Put `calllike` first to bail early

File: src/services/services.ts [hunk 5]
```diff
@@ -2788,20 +2788,36 @@ namespace ts {
...
-        }
+        node = climbPastPropertyAccess(node);
         return node && node.parent && node.parent.kind === SyntaxKind.NewExpression && (<CallExpression>node.parent).expression === node;
     }
 
+    /** Returns a CallLikeExpression where `node` is the target being invoked. */
+    function getAncestorCallLikeExpression(node: Node): CallLikeExpression | undefined {
+        const target = climbPastManyPropertyAccesses(node);
+        const callLike = target.parent;
+        return callLike && isCallLikeExpression(callLike) && getInvokedExpression(callLike) === target && callLike;
```
Comment (zhengbli) [inline src/services/services.ts]: Seems you are checking `callLike` twice

File: src/services/services.ts [hunk 6]
```diff
@@ -5068,14 +5084,25 @@ namespace ts {
...
 
+        function getSymbolInfo(typeChecker: TypeChecker, symbol: Symbol, node: Node) {
+            return {
+                symbolName: typeChecker.symbolToString(symbol), // Do not get scoped name, just the name of the symbol
+                symbolKind: getSymbolKind(symbol, node),
+                containerName: symbol.parent ? typeChecker.symbolToString(symbol.parent, node) : ""
+            };
+        }
+
+        function getDefinitionFromSignatureDeclaration(decl: SignatureDeclaration): DefinitionInfo {
```
Comment (zhengbli) [inline src/services/services.ts]: `createDefinitionFromSignatureDeclaration`

**Comment by zhengbli** [comment]
Other than the comments, :+1:

File: src/services/services.ts (hunk 5 above)
Comment (andy-ms) [inline src/services/services.ts]: First one is checking whether it exists. Last one is returning it from the function.

**Pull request merged**